tests/HttpClient/HttpClientFactoryTest.php

```diff
@@ -101,4 +116,79 @@ public function testCreateThrowsIfHttpProxyOptionIsUsedWithCustomHttpClient(): v
             'http_proxy' => 'http://example.com',
         ]));
     }
+
+    /**
+     * @runInSeparateProcess
+     */
+    public function testResolveClientWithSymfonyClient()
+    {
+        new SymfonyHttplugClient(SymfonyHttpClient::create([]));
+        $this->assertClientInstance(SymfonyHttplugClient::class);
+    }
+
+    /**
+     * @runInSeparateProcess
+     */
+    public function testResolveClientWithGuzzleClient()
+    {
+        GuzzleHttpClient::createWithConfig([]);
+        class_exists(RequestOptions::class);
+        $this->assertClientInstance(GuzzleHttpClient::class);
+    }
+
+    /**
+     * @runInSeparateProcess
+     */
+    public function testResolveClientWithCurlClient()
+    {
+        new CurlHttpClient();
+        $this->assertClientInstance(CurlHttpClient::class);
+    }
+
+    /**
+     * @runInSeparateProcess
+     */
+    public function testResolveClientWithDefaultClient()
+    {
+        class_exists(HttpAsyncClientDiscovery::class);
+        $mock = $this->createMock(HttpAsyncClientInterface::class);
+        $reflectedClass = new \ReflectionClass(ClassDiscovery::class);
+        $reflectedClass->setStaticPropertyValue('cache', [HttpAsyncClientInterface::class => ['class' => \get_class($mock)]]);
+        $this->assertClientInstance(\get_class($mock));
+    }
+
+    private function assertClientInstance(string $expectedClientClass)
+    {
+        $httpClientFactory = new HttpClientFactory(
+            Psr17FactoryDiscovery::findUrlFactory(),
+            Psr17FactoryDiscovery::findResponseFactory(),
+            Psr17FactoryDiscovery::findStreamFactory(),
+            null,
+            'sentry.php.test',
+            '1.2.3'
+        );
+        class_exists(HeaderSetPlugin::class);
```

Comment:
Correct me if I'm wrong, but I think that with all these `class_exists` function calls you are trying to assert that those classes got loaded by the HTTP client factory. However, I see at least two problems here: the first is that the first time such classes are loaded is when calling the `HttpClientFactory::create()` method and not its constructor, and the second is that the default value of the `$autoload` parameter of `class_exists` is `true`, so there is no assurance that those classes have been loaded before. As a last thing, testing the internal state of a class is bad for a lot of reasons, we did it in the past and I would like to avoid it since than. I wonder if this class has too many responsibilities and if a different approach would benefit both the maintenance and the tests (e.g. having a factory for each type of HTTP client we support)